daiquiri/__init__.py

```diff
@@ -51,21 +51,66 @@ def process(self, msg, kwargs):
         return msg, kwargs
 
 
+PARAMETER_TEMPLATE = ' [{0}: {1}]'
+
+
+class ArbitraryContextAdapter(logging.LoggerAdapter):
+    """Logger adapter to add context string to log record's extra data
+
+    Keywords passed to the log call are formatted into a context
+    string, which is then added to the "extra" dictionary passed to
+    the underlying logger so they are emitted with the log message and
+    available to the format string.
+
+    Example:
+      A formatter with the format string of "%(context)s %(messages)"
+      called like logger.info("A message.", test1="a", test2="b")
+      would log the message: " [test1: a] [test2: b] A message."
+
+    Special keywords:
+
+    extra
+      An existing dictionary of extra values to be passed to the
+      logger. If present, the dictionary is copied and extended.
+    """
+
+    def process(self, msg, kwargs):
+        # Make a new extra dictionary combining the values we were
+        # given when we were constructed and a 'context' key which
+        # will contain a formatted string containing anything from
+        # kwargs.
+        extra = self.extra.copy()
+        if 'extra' in kwargs:
+            extra.update(kwargs.pop('extra'))
+        # Format any unknown keyword arguments into the context string.
+        context = ''
+        for name in list(kwargs.keys()):
+            if name == 'exc_info':
+                continue
+            context += PARAMETER_TEMPLATE.format(name, kwargs.pop(name))
+        extra['context'] = context
+        extra['_daiquiri_extra'] = extra
+        kwargs['extra'] = extra
+        return msg, kwargs
```

Comment:
Haha I definitely understand you're not trying to annoy me :)

One other solution I can think of that might work to add functionality with minimal confusion would be to not add a new LoggingAdapter, but rather add the functionality within `ColorFormatter.format()`. Basically if a user wanted to use the `%(rest-of-extra)` key they would have pass an optional parameter to the ColorFormatter instantiation containing all of the keys they've specified in the format string. Then `ColorFormatter.format()` could filter those keys out of the `extra` dict and format the remainder into a string, add that onto the `log_record.rest-of-extra` attribute, and let the call to `super.format()` format the string. From the user's perspective I think it'd look like:

```python
>>> name = "my_module"
>>> logger = daiquiri.getLogger(name, level=logging.INFO)
>>> handler = logging.StreamHandler(sys.stdout)
>>> format_string = "%(asctime)s %(name)s %(levelname)s %(mood)s %(context)s: %(message)"
>>> handler.setFormatter(daiquiri.formatter.ColorFormatter(fmt=format_string,
                                                           keywords=["mood"]))
>>> logger.logger.addHandler(handler)
>>> logger.info("A message.", mood="happy", test1="a", test2="b")
2017-08-18 17:10:20,212 my_module INFO happy [test1: a] [test2: b]: A message.
>>> logger.info("Another message.", mood="sad", test1="a", test2="b", test3="c")
2017-08-18 17:10:20,212 my_module INFO sad [test1: a] [test2: b] [test3: c]: A message.
```

We can limit when the `rest-of-extra` functionality is "turned on" to only when the formatter's `keywords` attribute is set. This approach has the advantage that no new classes are required (and therefore don't need to be specified), a user not interested in the `rest-of-extra` functionality doesn't need to change anything, and a user who is interested in that functionality is able to do so by just dropping `%(rest-of-extra)` into their format string and specifying the `keywords` attribute on the formatter.